core/src/solve.jl

```diff
@@ -4,6 +4,40 @@ const ScalarInterpolation =
 const VectorInterpolation =
     LinearInterpolation{Vector{Vector{Float64}}, Vector{Float64}, true, Vector{Float64}}
 
+struct SparseMatrixCSC_DiffCache{Ti <: Integer, D <: DiffCache}
+    m::Int                  # Number of rows
+    n::Int                  # Number of columns
+    colptr::Vector{Ti}      # Column j is in colptr[j]:(colptr[j+1]-1)
+    rowval::Vector{Ti}      # Row indices of stored values
+    nzval::D                # Stored values, typically nonzeros
+end
+
+struct SparseMatrixCSC_cache{Ti <: Integer, D} <:
+       SparseArrays.AbstractSparseMatrixCSC{eltype(D), Ti}
+    m::Int                  # Number of rows
+    n::Int                  # Number of columns
+    colptr::Vector{Ti}      # Column j is in colptr[j]:(colptr[j+1]-1)
+    rowval::Vector{Ti}      # Row indices of stored values
+    nzval::D                # Stored values, typically nonzeros
+end
+
+const SparseCache = Union{SparseMatrixCSC_DiffCache, SparseMatrixCSC_cache}
+
+SparseArrays.size(matrix::SparseCache) = (matrix.m, matrix.n)
+SparseArrays.getcolptr(matrix::SparseCache) = matrix.colptr
+SparseArrays.rowvals(matrix::SparseCache) = matrix.rowval
+SparseArrays.nonzeros(matrix::SparseCache) = matrix.nzval
```

Comment:
Could you explain why you need to implement your own AbstractSparseMatrixCSC types? It seems the only difference is the `eltype`. That is the first type parameter:

```
SparseMatrixCSC{Tv,Ti<:Integer} <: AbstractSparseMatrixCSC{Tv,Ti}
```

Is it also possible to define separate methods for different `Tv`, to avoid needing our own sparse matrix types?